### PR TITLE
run pylint and pep8 quality tasks

### PR DESCRIPTION
**Pull request by muhammad-ammar**
This will add pylint and pep8 quality tasks. These tasks will run with every build. For now we have defined below violation limits.

PEP8_THRESHOLD = 50
PYLINT_THRESHOLD = 847

Build will fail if a PR has violations greater then the above limits.

@muzaffaryousaf @Qubad786 @mushtaqak @yro

### CODE REVIEW

File: control/tests/test_val.py [hunk 1]
```diff
@@ -10,8 +10,8 @@
 import responses
 
 from control.veda_val import VALAPICall
-from veda_file_ingest import VideoProto
-from VEDA_OS01 import utils
+from control.veda_file_ingest import VideoProto
+from VEDA import utils
```
Comment (Qubad786) [inline control/tests/test_val.py]: A test is failing on this one.

**Comment by muzaffaryousaf** [comment]
@muhammad-ammar I think we should add quality build as well in this PR.

**Comment by muzaffaryousaf** [comment]
@muhammad-ammar let's finalize & merge this one too.

File: pavelib/quality.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    flags = []
+    if errors:
+        flags.append("--errors-only")
+
+    sh(
+        "PYTHONPATH={python_path} pylint {packages} {flags} --msg-template={msg_template} | "
+        "tee {report_dir}/pylint.report".format(
+            python_path=ROOT_DIR,
+            packages=' '.join(PACKAGES),
+            flags=" ".join(flags),
```
Comment (Qubad786) [inline pavelib/quality.py]: single quotes.

File: pavelib/quality.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            report_dir=REPORTS_DIR
+        )
+    )
+
+    num_violations += _count_pylint_violations(
+        "{report_dir}/pylint.report".format(report_dir=REPORTS_DIR)
+    )
+
+    # Print number of violations to log
+    violations_count_str = "Number of pylint violations: " + str(num_violations)
```
Comment (Qubad786) [inline pavelib/quality.py]: use `format` instead.

File: pavelib/quality.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    print violations_count_str
```
Comment (Qubad786) [inline pavelib/quality.py]: Do we need to print?

File: pavelib/quality.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    pylint_pattern = re.compile(r".(\d+):\ \[(\D\d+.+\]).")
+
+    for line in open(report_file):
+        violation_list_for_line = pylint_pattern.split(line)
+        # If the string is parsed into four parts, then we've found a violation. Example of split parts:
+        # test file, line number, violation name, violation details
+        if len(violation_list_for_line) == 4:
+            num_violations_report += 1
+
+    return num_violations_report
```
Comment (mushtaqak) [inline pavelib/quality.py]: New line

File: pavelib/quality.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+@task
+@cmdopts([
+    ("limit=", "l", "limit for number of acceptable violations"),
+])
+def run_pep8(options):  # pylint: disable=unused-argument
+    """
+    Run pep8 on system code.
+    Fail the task if any violations are found.
+    """
+    violations_limit = int(getattr(options, 'limit', -1))
```
Comment (mushtaqak) [inline pavelib/quality.py]: why -1?

File: pavelib/quality.py (hunk 3 above)
Comment (mushtaqak) [inline pavelib/quality.py]: We can use `str.format` here

Comment (mushtaqak) [reply pavelib/quality.py]: No need to store variable, we can just print this?

File: pavelib/quality.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    violations_message = "Too many pep8 violations. Number of pep8 violations: {num_violations}. ".format(
+        num_violations=num_violations
+    )
+    violations_message += "The limit is {violations_limit}. ".format(violations_limit=violations_limit)
+    violations_message += "\n\nViolations:\n{violations_list}".format(violations_list=violations_list)
+
+    # Fail if number of violations is greater than the limit
+    if num_violations > violations_limit > -1:
+        raise BuildFailure(violations_message)
```
Comment (mushtaqak) [inline pavelib/quality.py]: I think we can create `violations_message` here

File: pavelib/quality.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    violations_limit = int(getattr(options, 'limit', -1))
+
+    sh('pep8 . | tee {report_dir}/pep8.report'.format(report_dir=REPORTS_DIR))
+
+    num_violations, violations_list = _pep8_violations(
+        "{report_dir}/pep8.report".format(report_dir=REPORTS_DIR)
+    )
+    violations_list = ''.join(violations_list)
+
+    violations_message = "Too many pep8 violations. Number of pep8 violations: {num_violations}. ".format(
```
Comment (mushtaqak) [inline pavelib/quality.py]: Better to create `violations_message` inside if block

File: pavelib/quality.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        "tee {report_dir}/pylint.report".format(
+            python_path=ROOT_DIR,
+            packages=' '.join(PACKAGES),
+            flags=" ".join(flags),
+            msg_template='"{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}"',
+            report_dir=REPORTS_DIR
+        )
+    )
+
+    num_violations += _count_pylint_violations(
```
Comment (Qubad786) [inline pavelib/quality.py]: its not in any loop can we just assign except `+=`?

File: pavelib/quality.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        "{report_dir}/pylint.report".format(report_dir=REPORTS_DIR)
+    )
+
+    # Print number of violations to log
+    violations_count_str = "Number of pylint violations: " + str(num_violations)
+    print violations_count_str
+
+    # Fail if number of violations is greater than the limit
+    if num_violations > violations_limit > -1:
+        raise BuildFailure("Failed. Too many pylint violations. "
```
Comment (mushtaqak) [inline pavelib/quality.py]: We are not showing violations in case of pylint?

File: pavelib/quality.py (hunk 5 above)
Comment (mushtaqak) [inline pavelib/quality.py]: New line

File: pavelib/quality.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+REPORTS_DIR = os.path.join(ROOT_DIR, 'reports')
+
+PACKAGES = [
+    'VEDA',
+    'VEDA_OS01',
+    'control',
+    'frontend',
+    'youtube_callback',
+    'scripts',
```
Comment (muzaffaryousaf) [inline pavelib/quality.py]: Can we explicitly include bin as well ?

File: pavelib/quality.py [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    )
+    violations_message += "The limit is {violations_limit}. ".format(violations_limit=violations_limit)
+    violations_message += "\n\nViolations:\n{violations_list}".format(violations_list=violations_list)
+
+    # Fail if number of violations is greater than the limit
+    if num_violations > violations_limit > -1:
+        raise BuildFailure(violations_message)
+
+
+def _pep8_violations(report_file):
```
Comment (mushtaqak) [inline pavelib/quality.py]: this can be rename to `_count_pep8_violations` so as to be consistent with pylint method

File: pavelib/quality.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                           "The limit is {violations_limit}.".format(violations_limit=violations_limit))
+
+
+def _count_pylint_violations(report_file):
+    """
+    Parses a pylint report line-by-line and determines the number of violations reported
+    """
+    num_violations_report = 0
+    # An example string:
+    # common/lib/xmodule/xmodule/tests/test_conditional.py:21: [C0111(missing-docstring), DummySystem] Missing docstring
```
Comment (mushtaqak) [inline pavelib/quality.py]: Are we going to add this example here ?

Comment (muzaffaryousaf) [inline pavelib/quality.py]: do we need to update this comment ?

**Comment by muhammad-ammar** [comment]
@muzaffaryousaf @Qubad786 @mushtaqak Feedback addressed.

File: pavelib/quality.py (hunk 11 above)
Comment (muzaffaryousaf) [reply pavelib/quality.py]: This is still not being addressed.

**Pull request merged**